PyInstaller/hooks/rthooks/pyi_rth_pkgres.py

```diff
@@ -9,14 +9,179 @@
+
+        return current
+
+    def path_exists(self, path):
+        node = self._get_tree_node(path)
+        return node is not None  # File or directory
+
+    def path_isdir(self, path):
+        node = self._get_tree_node(path)
+
+        if node is None:
+            return False  # Non-existant
+
+        if isinstance(node, str):
+            return False  # File
+
+        return True
+
+    def path_listdir(self, path):
+        node = self._get_tree_node(path)
+
+        if not isinstance(node, dict):
+            return []  # Non-existant or file
+
+        return list(node.keys())
+
+
+class PyiFrozenProvider(pkg_resources.NullProvider):
+    def __init__(self, module):
+        super().__init__(module)
+
+        # Get top-level path; if "module" corresponds to a package,
+        # we need the path to the package itself. If "module" is a
+        # submodule in a package, we need the path to the parent
+        # package.
+        self._pkg_path = pathlib.PurePath(module.__file__).parent
+
+        # Construct a path relative to _MEIPASS directory for searching
+        # the TOC
+        rel_pkg_path = self._pkg_path.relative_to(SYS_PREFIX)
+
+        # Reconstruct package name prefix (use package path to obtain
+        # correct prefix in case of a module)
+        pkg_name = '.'.join(rel_pkg_path.parts)
+
+        # Collect relevant entries from TOC. We are interested in either
+        # files that are located in the package/module's directory (data
+        # files) or in packages that are prefixed with package/module's
+        # name (to reconstruct subpackage directories)
+        data_files = []
+        package_dirs = []
+
+        for entry in self.loader.toc:
+            entry_path = pathlib.PurePath(entry)
+
+            if rel_pkg_path in entry_path.parents:
+                # Data file path
+                data_files.append(entry_path)
+            elif entry.startswith(pkg_name) and self.loader.is_package(entry):
+                # Package or subpackage; convert the name to directory path
+                package_dir = pathlib.PurePath(*entry.split('.'))
+                package_dirs.append(package_dir)
+
+        # Reconstruct the filesystem
+        self.embedded_tree = _TocFilesystem(data_files, package_dirs)
+
+    def _normalize_path(self, path):
+        # Avoid using Path.resolve(), because it resolves symlinks. This
+        # is undesirable, because the pure path in self._pkg_path does
+        # not have symlinks resolved, so comparison between the two
+        # would be faulty. So use os.path.abspath() instead to normalize
+        # the path
+        return pathlib.Path(os.path.abspath(path))
+
+    def _is_relative_to_package(self, path):
+        return path == self._pkg_path or self._pkg_path in path.parents
+
+    def _has(self, path):
+        # Prevent access outside the package
+        path = self._normalize_path(path)
+        if not self._is_relative_to_package(path):
```

Comment:
The `DefaultProvider` currently allows querying/listing of paths outside the given module/package via `..`, with an added deprecation warning: `<stdin>:1: DeprecationWarning: Use of .. or absolute path in a resource path is not allowed and will raise exceptions in a future release.`. 

`ZipProvider` cannot do it, and thus always returns `False` for `_has()` or `_isdir()`, and empty list for `_listdir()`.

In our `PyFrozenImplementation`, we explicitly disallow the jump, to keep behavior consistent between the PYZ-embedded and on-filesystem resources. In the filesystem case, the parent path could be queried otherwise. But in PYZ embeded case, the current implementation cannot query the parent path - even if the path was still within the "main" package (e.g., parent to a sub-package but still within the main one) because the prefix tree is constructed only from the base (sub)package down.

The `pyi_pkg_resources_provider.py` does contain tests for trying to query paths outside the base module/package to validate the behavior of `DefaultProvider`, `NullProvider`, and `PyiFrozenProvider`.